File: src/main/java/com/web/rpg/service/monster/impl/MonsterServiceImpl.java
package com.web.rpg.service.monster.impl;

import com.web.rpg.model.Characters.CharacterClass;
import com.web.rpg.model.Characters.PlayerCharacter;
import com.web.rpg.model.Monsters.Monster;
import com.web.rpg.repository.MonsterRepository;
import com.web.rpg.service.monster.MonsterService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.util.List;
import java.util.Properties;
import java.util.Random;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class MonsterServiceImpl implements MonsterService {

    @Autowired
    private MonsterRepository monsterRepository;

    private static final Random RANDOM = new Random();

    @Override
    public void applyDamage(Monster monster, double damage) {
        double currentHitPoint = monster.getHitPoint() - damage;
        if (currentHitPoint >= 0) {
            monster.setHitPoint(currentHitPoint);
        } else {
            monster.setHitPoint(0);
        }
    }

    @Override
    public Monster prepearMonsterForBattle(PlayerCharacter character){
        Monster monster = new Monster();
        monster.setMonsterId(UUID.randomUUID());
        if (CharacterClass.BERSERK.equals(character.getCharacterClass())) {
            monster.setLevel(character.getLevel() + 4);
        } else {
            monster.setLevel(character.getLevel() + 1);
        }
        if (character.getLevel() > 5){
            monster.setExperience(character.getLevel() * 50);
            monster.setHitPoint(monster.getLevel() * 150);
            monster.setDamage(monster.getLevel() * 40);
        }else {
            monster.setExperience(character.getLevel()*10);
            monster.setHitPoint(monster.getLevel() * 70);
            monster.setDamage(monster.getLevel() * 20);
        }
        monster.setName(generateName());
        return monster;
    }

    @Override
    public void remove(Monster monster) {
        monsterRepository.delete(monster);
    }

    @Override
    public Monster updateOrCreate(Monster monster) {
        monsterRepository.save(monster);
        return monsterRepository.findOne(((Monster)monster).getMonsterId());
    }

    private String generateName() {
        Properties prop = new Properties();
        try {
            prop.load(getClass().getClassLoader().getResourceAsStream("properties/monster-names.properties"));
        } catch (IOException e) {
            e.printStackTrace();
        }
        List<String> characterNames = prop.entrySet()
                .stream()
                .map(entry -> (String) entry.getValue())
                .collect(Collectors.toList());
        return characterNames.get(RANDOM.nextInt(characterNames.size()));
    }
}

File: src/main/java/com/web/rpg/service/monster/impl/MonsterServiceImpl.scala
package com.web.rpg.service.monster.impl

import java.io.IOException
import java.util
import java.util.{Properties, Random, UUID}

import scala.collection.JavaConverters._
import com.web.rpg.model.Characters.{CharacterClass, PlayerCharacter}
import com.web.rpg.model.Monsters.Monster
import com.web.rpg.repository.MonsterRepository
import com.web.rpg.service.monster.MonsterService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

import scala.collection.mutable

@Service
class MonsterServiceImpl @Autowired()(val monsterRepository: MonsterRepository) extends MonsterService {

  private val RANDOM: Random = new Random

  override def applyDamage(monster: Monster, damage: Double): Unit = {
    val currentHitPoint: Double = monster.getHitPoint - damage
    if (currentHitPoint >= 0) monster.setHitPoint(currentHitPoint)
    else monster.setHitPoint(0)
  }

  override def prepareMonsterForBattle(character: PlayerCharacter): Monster = {
    val monster: Monster = new Monster
    monster.setMonsterId(UUID.randomUUID)
    if (CharacterClass.BERSERK == character.getCharacterClass) {
      monster.setLevel(character.getLevel + 4)
    } else {
      monster.setLevel(character.getLevel + 1)
    }
    if (isNotBeginner(character)) {
      monster.setExperience(character.getLevel * 50)
      monster.setHitPoint(monster.getLevel * 150)
      monster.setDamage(monster.getLevel * 40)
    } else {
      monster.setExperience(character.getLevel * 10)
      monster.setHitPoint(monster.getLevel * 70)
      monster.setDamage(monster.getLevel * 20)
    }
    monster.setName(loadName)
    monster
  }

  override def remove(monster: Monster): Unit = {
    monsterRepository.delete(monster)
  }

  override def updateOrCreate(monster: Monster): Monster = {
    monsterRepository.save(monster)
    monsterRepository.findOne(monster.asInstanceOf[Monster].getMonsterId)
  }

  private def loadName: String = {
    val prop: Properties = new Properties
    try
      prop.load(getClass.getClassLoader.getResourceAsStream("properties/monster-names.properties"))
    catch {
      case e: IOException =>
        e.printStackTrace()
    }

    val characterNames: mutable.Set[String] = prop.entrySet()
      .asScala
      .map((entry: util.Map.Entry[AnyRef, AnyRef]) => entry.getValue.toString)
      .collect{case name: String => name}
    characterNames.toList(RANDOM.nextInt(characterNames.size))
  }

  private def isNotBeginner(character: PlayerCharacter) = {
    character.getLevel > 5
  }
}
